crates/amaru-ledger/src/rules/transaction/scripts.rs

```diff
@@ -1,59 +1,115 @@
+            context
+                // We assume that the input exists as that's validated during the inputs validation
+                .lookup(input)
+                .map(|output| (input, output))
+        })
+        .collect::<Vec<_>>();
+
+    let empty_vec = vec![];
+    let resolved_reference_inputs = reference_inputs
+        .unwrap_or(&empty_vec)
+        .iter()
+        .filter_map(|input| {
+            context
+                // We assume that the reference input exists as that's validated during the inputs validation
+                .lookup(input)
+                .map(|output| (input, output))
+        })
+        .collect::<Vec<_>>();
+
     // provided reference scripts from inputs and reference inputs only include ScriptRefs that are required by an input
     let provided_reference_scripts = [
-        reference_inputs.unwrap_or(&vec![]).as_slice(),
-        inputs.as_slice(),
+        resolved_inputs.as_slice(),
+        resolved_reference_inputs.as_slice(),
     ]
     .concat()
     .iter()
-    .filter_map(|input| {
-        context
-            // We assume that the reference input exists as that's validated during the inputs validation
-            .lookup(input)
-            .and_then(|output| match output {
-                amaru_kernel::PseudoTransactionOutput::PostAlonzo(transaction_output) => {
-                    transaction_output
-                        .script_ref
-                        .as_deref()
-                        .and_then(|script_ref| {
-                            // If there is a provided ScriptRef, make sure it is required by an input
-                            let hash = script_ref.script_hash();
-                            if required_scripts.contains(&hash) {
-                                Some(hash)
-                            } else {
-                                None
-                            }
-                        })
-                }
-                amaru_kernel::PseudoTransactionOutput::Legacy(_) => None,
-            })
+    .filter_map(|(_, output)| {
+        match output {
+            amaru_kernel::PseudoTransactionOutput::PostAlonzo(transaction_output) => {
+                transaction_output
+                    .script_ref
+                    .as_deref()
+                    .and_then(|script_ref| {
+                        // If there is a provided ScriptRef, make sure it is required by an input
+                        let hash = script_ref.script_hash();
+                        if required_scripts.contains(&hash) {
+                            Some(ScriptRefWithHash {
+                                hash: script_ref.script_hash(),
+                                script: script_ref.clone(),
+                            })
+                        } else {
+                            None
+                        }
+                    })
+            }
+            amaru_kernel::PseudoTransactionOutput::Legacy(_) => None,
```

Comment:
I know it's just re-indenting, but it now occurs to me that this feels like a wrong abstraction level here. I would hope that the ledger rules can be free of serialisation details (i.e. we shall never have to pattern match on a `PostAlonzo` struct in here). 

So there's room for moving stuff out as helper function or trait.

Reply:
I agree with this. I'll write up a trait